Add video tracks functionality

jorgefilipecosta opened the pull request:
Implements tracks adding and edition functionality on the video block.
Fixes: https://github.com/WordPress/gutenberg/issues/7673

This PR follows the design proposed by @enriquesanchez with some adaptations because of the way our components work.


## How has this been tested?
I added a video block, with a video file.
I used the tracks option on the toolbar.
I selected the upload option.
I verified a system window to select a file to upload opened.
I verified this window only allow to pick .vtt (all tracks on the tracks element must be in the vtt format).
I uploaded the file.
I saved the post and verified the track worked as expected.
I added another video block and used the tracks option.
I selected the "Open Media Library Option" and verified the media library only showed vtt files.
I verified I could edit the tracks.
I verified I could remove the tracks.

## Screenshots <!-- if applicable -->

![image](https://user-images.githubusercontent.com/11271197/95207122-6cd60980-07df-11eb-9f2a-675242371b85.png)

![image](https://user-images.githubusercontent.com/11271197/95207191-80817000-07df-11eb-9b10-d885919bc09c.png)

jorgefilipecosta committed:
Add video tracks functionality

enriquesanchez commented:
Thank you, @jorgefilipecosta! This is looking really really good.

I have some feedback on the design:

1. We need a `tooltip` and `aria-label` for the tracks button in the block toolbar. It should say 'Text tracks'.
2. The `menu group label` should also say 'Text tracks'.
3. When selecting to upload a track file, keyboard focus should be placed on the first focusable element, the `label` field.
4. When uploading a track, the 'Ok' button should say 'Add track' instead.
5. Let's add a 'Cancel' button next to it. This should let users cancel the process in case they change their mind.
6. I also think we should not show the 'Remove' button when the user is in the upload process. Removing should only be allowed once a track has been added.
7. When editing a track, the 'Ok' button should say 'Save' and we should also have the option to 'Cancel'.
8. When editing a track, we shuould display the 'Remove track' button.

Do you have any context on why we only support the `vtt` format? I know `srt` is a very common format for subtitles and I wonder why that one is not allowed.

Again, thank you for working on this 🎉

jorgefilipecosta commented:
Hi @enriquesanchez, thank you for the review. 👍
I will update this PR to address your feedback.

> Let's add a 'Cancel' button next to it. This should let users cancel the process in case they change their mind.
> Let's add a 'Cancel' button next to it. This should let users cancel the process in case they change their mind.
> I also think we should not show the 'Remove' button when the user is in the upload process. Removing should only be allowed once a track has been added.
When editing a track, the 'Ok' button should say 'Save' and we should also have the option to 'Cancel'.


I did not add a Save because a save would not be reflecting what is happening. When the user changes something, e.g.: the label the attribute is already saved, there would be nothing to save. The changes are immediate and are saved on each keypress. That's what happens on all the other cases on the block editor, e.g., inspector changes; if we change a field, the change is applied right away. The user, of course, can use undo to undo the change. What the ok is doing is just going back to the tracklist. It is not saving anything. That's also the reason why we don't have a cancel because the cancel would also just go back to the tracklist. So Ok and Cancel are the same thing given how things are implemented.
I tried to follow the general philosophy of saving the attributes right away. I think we don't have other cases with a save operation.

But if we think in this case, we have a special situation and should not follow the pattern of the immediate block attribute changes, I can implement a local state on the component. User changes are saved to the local state. On the save operation, the changes are saved to the attributes, and on the cancel, they are discarded, and the local state is reset.
In that case, undo/redo would not affect local state changes. E.g.: the user changes the label and uses undo before pressing save, undo would affect other action, and the label the user wrote would still be there (the change would be discarded with cancel).
This situation is, in fact a little different from all the other cases, so It may make sense to implement something special, let me know what you think.


> Do you have any context on why we only support the vtt format? I know srt is a very common format for subtitles and I wonder why that one is not allowed.

The track element specification says only vtt (Web Video Text Tracks) format for the track element as we can check at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track.

jorgefilipecosta commented:
> We need a tooltip and aria-label for the tracks button in the block toolbar. It should say 'Text tracks'.
> The menu group label should also say 'Text tracks'.
> When selecting to upload a track file, keyboard focus should be placed on the first focusable element, the label field.

enriquesanchez commented:
>I did not add a Save because a save would not be reflecting what is happening. 

Thank you for that detailed clarification, @jorgefilipecosta. I don't think this should be a special case and perhaps the design needs a bit of adjusting so we don't mislead users.

The presence of a Save/Done/OK button does create the illusion of there being a confirmation step when in reality there isn't one. That said, because this is not a simple menu popover, I think there should still be a clear way to close it. In light of this, I think that your initial implementation is right; we just need to change the button to use the default style and change the label to 'Close' so it's more clear what is happening:

<img width="360" alt="Frame 33" src="https://user-images.githubusercontent.com/3276087/95374560-1aaaec00-08a4-11eb-8d0b-d96f1f469fd2.png">

How does that feel to you?

---

I also just noticed that in the PR the file name is missing in the 'File:'  field up top:

<img width="388" alt="Screen Shot 2020-10-07 at 14 08 26" src="https://user-images.githubusercontent.com/3276087/95376406-a887d680-08a6-11eb-83c1-fc240e4ee19f.png">

mtias commented:
Can we add a little inline description on what "tracks" are? It's can be quite inscrutable as is.

jorgefilipecosta commented:
Hi @mtias, 
Before any track is added there is a small text indicating what a track is:
![image](https://user-images.githubusercontent.com/11271197/95609976-7d90b480-0a57-11eb-9647-b11861e3157c.png)

Would you like that this text is expended or would want to show this message in other states e.g: there are tracks already added or the user is editing a track?

jorgefilipecosta commented:
Hi @enriquesanchez,
> I also just noticed that in the PR the file name is missing in the 'File:' field up top:

It was fixed.

Your feedback was applied and the UI updates were implemented.

mtias commented:
> Would you like that this text is expended or would want to show this message in other states e.g: there are tracks already added or the user is editing a track?

Ah, thanks. I think it can make sense to keep it always displayed, but let's see how it feels

enriquesanchez commented:
@jorgefilipecosta Thanks for the updates 🙌 

I'm still not seeing the file name displayed. I made sure to update the branch and tried in both Firefox and Safari.

<img width="628" alt="Screen Shot 2020-10-12 at 16 09 48" src="https://user-images.githubusercontent.com/3276087/95791211-80351900-0ca6-11eb-9560-54bfdff0a524.png">

---

I also tested with VoiceOver and I think we can add a small improvement to make it more accessible. The 'Edit' button for each text track should say what text track you'll be editing. I think we can leave the 'Edit' label as is and just add an `aria-label` that is more descriptive: 'Edit [title of track]' (_ie: Edit French subtitles_).

I also noticed that you can leave these fields empty, which results in this:

<img width="628" alt="Screen Shot 2020-10-12 at 16 25 39" src="https://user-images.githubusercontent.com/3276087/95791795-b626cd00-0ca7-11eb-87f0-3b9cc0b5b3c7.png">

I don't think we should allow for this to happen and should make these fields required. Alternativeky, we can at least display some defaults. For example, if no label is provided we can use the file name as the track's label. 

For reference the Classic editor defaults to `label="English",` `srclang="eb"`, and `kind="subtitles"`.

What do you think?

jorgefilipecosta commented:
> > Would you like that this text is expended or would want to show this message in other states e.g: there are tracks already added or the user is editing a track?
> 
> Ah, thanks. I think it can make sense to keep it always displayed, but let's see how it feels

I followed this suggestion by @mtias and added the text to the tracklist too:

![image](https://user-images.githubusercontent.com/11271197/95873312-b5ee0680-0d67-11eb-87f8-8670c4c3efaf.png)


> I also tested with VoiceOver and I think we can add a small improvement to make it more accessible. The 'Edit' button for each text track should say what text track is that you'll be editing. I think we can leave the 'Edit' label as is and just add an aria-label that is more descriptive: 'Edit [title of track]' (ie: Edit French subtitles).

Suggestion applied.


> I don't think we should allow for this to happen and should make these fields required. Alternatively, we can at least display some defaults. For example, if no label is provided we can use the file name as the track's label.

I followed the same approach the classic editor does. At the start the inputs are empty but if the user does not fill them they get the default values of "English" and "en" exactly what the classic editor does.


> I'm still not seeing the file name displayed. I made sure to update the branch and tried in both Firefox and Safari. Maybe I'm missing something?

I'm not being able to reproduce the issue:

![image](https://user-images.githubusercontent.com/11271197/95873800-26952300-0d68-11eb-8dae-c460df8cef8a.png)

Right after I upload a vtt file for like 1 second the field is empty but after the file is uploaded the name appears, and each time I go to edit.
Would you be able to share the steps you did, the vtt file you are using, and the markup of the block (switch to the code editor and copy paster what is there)? Thank you in advance this may be helpful for me to understand where is the bug.

enriquesanchez commented:
@mtias Would like for us to reconsider showing the description of what text tracks are all the time. I think it's enough to display it when there are no text tracks/ the first time the menu is opened. Authors will encounter the description every time they add a new video, for example.

I don't think it's necesssary to show it after tracks have been added. By then the user should already be familiar with the feature, and not having it helps us keep the menu more condensed in the case multiple tracks are added.

enriquesanchez commented:
Thank you for working on these updates, @jorgefilipecosta. This is in much better shape now. 🙌 

I noticed a small issue when navigating with keyboard. When opening the menu, if you use the down arrow the menu closes and focus is lost and moved up to the title of the document. I expect that users might try to use up/down arrow keys when they open this menu for the first time because up/down is used to navigate other menus in the block editor. We should make sure that focus is retained inside the popover.

youknowriad commented on packages/block-library/src/video/tracks-editor.js:
```diff
@@ -0,0 +1,305 @@
+}
+
+export default function TracksEditor( { tracks = [], onChange } ) {
+	const mediaUpload = useSelect( ( select ) => {
```
Just noting that this can be undefined, in which case we should probably hide the button entirely right?

youknowriad replied:
Or maybe allow using external  URLs (can be for later).

jorgefilipecosta replied:
Yah not rendering anything seems like a good option for now. I updated the code.

youknowriad commented on packages/block-library/src/video/edit.js:
```diff
@@ -145,6 +147,12 @@ function VideoEdit( {
+				<TracksEditor
+					tracks={ tracks }
+					onChange={ ( newTracks ) => {
+						setAttributes( { tracks: newTracks } );
```
Just want to ask whether the shape of the "track" object set here is well defined and that we're not dumping the whole object from the API or something like that.

jorgefilipecosta replied:
Yes, the shape is very specific an array of objects that may have {src, label, srcLang, kind } properties all being strings. We never dump anything from an API there we are intentional in storing the src explicitly (the other attributes came from user inputs).

mtias commented:
@enriquesanchez sure, don't have a strong opinion on it outside of the fact I think we can generally use contextual guidance built in into the UI better.

jorgefilipecosta commented:
Hi @enriquesanchez,

> I noticed a small issue when navigating with keyboard. When opening the menu, if you use the down arrow the menu closes and focus is lost and moved up to the title of the document. I expect that users might try to use up/down arrow keys when they open this menu for the first time because up/down is used to navigate other menus in the block editor. We should make sure that focus is retained inside the popover.

It was fixed.


I also reverted and made the help text only appear before the first track is added.

youknowriad reviewed (approved):
LGTM code wise

enriquesanchez commented:
Much better, thank you @jorgefilipecosta 👏 

I'm still finding that the popover closes and my focus moves up to the document title when I use up/down arrows in the edit screen. 

<img width="599" alt="Screen Shot 2020-10-14 at 16 46 33" src="https://user-images.githubusercontent.com/3276087/96048876-d936b580-0e3c-11eb-9196-da7aa2f0696e.png">

Can you reproduce this?

jorgefilipecosta commented:
> I'm still finding that the popover closes and my focus moves up to the document title when I use up/down arrows in the edit screen.

Nice catch @enriquesanchez the issue was fixed.

enriquesanchez commented:
👏 This is working really well, @jorgefilipecosta. Thank you!

I'm making some minor CSS tweaks on alignment and spacing of labels and fields and once I have that ready we should be good to go.

enriquesanchez committed:
Tweaks to alignment and spacing

jorgefilipecosta committed:
Remove hardcoded css margins

youknowriad commented on packages/block-library/src/video/style.scss:
```diff
@@ -102,3 +87,18 @@
 	padding: $grid-unit-15;
 }
+
+.block-library-video-tracks-editor__single-track-editor .components-base-control {
```
these look like editor specific styles so they should move to `editor.scss` right?

youknowriad replied:
Actually the whole changes on this file  should probably move there.

enriquesanchez replied:
I'm not sure how these changes will affect the whole of the editor. My intention was to specifically target the video tracks popover.

ockham commented on packages/block-library/src/video/tracks.js:
```diff
@@ -0,0 +1,5 @@
+export default function Tracks( { tracks = [] } ) {
+	return tracks.map( ( track ) => {
+		return <track key={ track.src } { ...track } />;
+	} );
```
Maybe we should use a different variable name in the loop that can't be confused with the `<track />` tag name easily?

```suggestion
	return tracks.map( ( t ) => {
		return <track key={ t.src } { ...t } />;
	} );
```
